lib/BoundaryValueDiffEqFIRK/src/firk.jl

```diff
@@ -441,43 +441,65 @@ function __construct_nlproblem(
     L = length(resid_bc)
     resid_collocation = __similar(y, cache.M * (N - 1) * (stage + 1))
 
-    loss_bcₚ = (iip ? __Fix3 : Base.Fix2)(loss_bc, cache.p)
-    loss_collocationₚ = (iip ? __Fix3 : Base.Fix2)(loss_collocation, cache.p)
+    bc_diffmode = if jac_alg.bc_diffmode isa AutoSparse
+        AutoSparse(jac_alg.bc_diffmode;
+            sparsity_detector = SparseConnectivityTracer.TracerSparsityDetector(),
+            coloring_algorithm = GreedyColoringAlgorithm(LargestFirst()))
+    else
+        jac_alg.bc_diffmode
+    end
 
-    sd_bc = jac_alg.bc_diffmode isa AutoSparse ? SymbolicsSparsityDetection() :
-            NoSparsityDetection()
-    cache_bc = __sparse_jacobian_cache(
-        Val(iip), jac_alg.bc_diffmode, sd_bc, loss_bcₚ, resid_bc, y)
+    cache_bc = if iip
+        DI.prepare_jacobian(loss_bc, resid_bc, bc_diffmode, y, Constant(cache.p))
+    else
+        DI.prepare_jacobian(loss_bc, bc_diffmode, y, Constant(cache.p))
+    end
 
-    sd_collocation = if jac_alg.nonbc_diffmode isa AutoSparse
+    nonbc_diffmode = if jac_alg.nonbc_diffmode isa AutoSparse
         if L < cache.M
             # For underdetermined problems we use sparse since we don't have banded qr
-            colored_matrix = __generate_sparse_jacobian_prototype(
-                cache, cache.problem_type, y, y, cache.M, N)
             J_full_band = nothing
-            __sparsity_detection_alg(ColoredMatrix(
-                sparse(colored_matrix.M), colored_matrix.row_colorvec,
-                colored_matrix.col_colorvec))
+            colored_result = __generate_sparse_jacobian_prototype(
+                cache, cache.problem_type, y, y, cache.M, N, jac_alg.nonbc_diffmode)
         else
             block_size = cache.M * (stage + 2)
             J_full_band = BandedMatrix(
                 Ones{eltype(y)}(L + cache.M * (stage + 1) * (N - 1),
                     cache.M * (stage + 1) * (N - 1) + cache.M),
                 (block_size, block_size))
-            __sparsity_detection_alg(__generate_sparse_jacobian_prototype(
-                cache, cache.problem_type, y, y, cache.M, N))
+            colored_result = __generate_sparse_jacobian_prototype(
+                cache, cache.problem_type, y, y, cache.M, N, jac_alg.nonbc_diffmode)
         end
+        AutoSparse(get_dense_ad(jac_alg.nonbc_diffmode);
+            sparsity_detector = ADTypes.KnownJacobianSparsityDetector(colored_result.A),
+            coloring_algorithm = ConstantColoringAlgorithm{ifelse(
+                ADTypes.mode(jac_alg.nonbc_diffmode) isa ADTypes.ReverseMode,
+                :row, :column)}(colored_result.A,
```

Comment:
```suggestion
                :row, :column)}(sparsity_pattern(colored_result),
```
Fields of the coloring result are private, you should use the accessors defined by SparseMatrixColorings.jl (here and everywhere else)

Reply:
Fixed